vendor/k8s.io/kubernetes/pkg/kubelet/kubelet.go

```diff
@@ -1148,78 +1151,149 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
+
 	for {
 		time.Sleep(step)
 		step = step * 2
 		if step >= 7*time.Second {
 			step = 7 * time.Second
 		}
 
-		node, err := kl.initialNodeStatus()
+		node, err := kl.initialNode()
 		if err != nil {
 			glog.Errorf("Unable to construct api.Node object for kubelet: %v", err)
 			continue
 		}
 
-		glog.V(2).Infof("Attempting to register node %s", node.Name)
-		if _, err := kl.kubeClient.Core().Nodes().Create(node); err != nil {
-			if !apierrors.IsAlreadyExists(err) {
-				glog.V(2).Infof("Unable to register %s with the apiserver: %v", node.Name, err)
-				continue
-			}
-			currentNode, err := kl.kubeClient.Core().Nodes().Get(kl.nodeName)
-			if err != nil {
-				glog.Errorf("error getting node %q: %v", kl.nodeName, err)
-				continue
-			}
-			if currentNode == nil {
-				glog.Errorf("no node instance returned for %q", kl.nodeName)
-				continue
+		glog.Infof("Attempting to register node %s", node.Name)
+		registered := kl.tryRegisterWithApiServer(node)
+		if registered {
+			glog.Infof("Successfully registered node %s", node.Name)
+			kl.registrationCompleted = true
+			return
+		}
+	}
+}
+
+// tryRegisterWithApiServer makes an attempt to register the given node with
+// the API server, returning a boolean indicating whether the attempt was
+// successful.  If a node with the same name already exists, it reconciles the
+// value of the annotation for controller-managed attach-detach of attachable
+// persistent volumes for the node.  If a node of the same name exists but has
+// a different externalID value, it attempts to delete that node so that a
+// later attempt can recreate it.
+func (kl *Kubelet) tryRegisterWithApiServer(node *api.Node) bool {
+	_, err := kl.kubeClient.Core().Nodes().Create(node)
+	if err == nil {
+		return true
+	}
+
+	if !apierrors.IsAlreadyExists(err) {
+		glog.Errorf("Unable to register node %q with API server: %v", kl.nodeName, err)
+		return false
+	}
+
+	existingNode, err := kl.kubeClient.Core().Nodes().Get(kl.nodeName)
+	if err != nil {
+		glog.Errorf("Unable to register node %q with API server: error getting existing node: %v", kl.nodeName, err)
+		return false
+	}
+	if existingNode == nil {
+		glog.Errorf("Unable to register node %q with API server: no node instance returned", kl.nodeName)
+		return false
+	}
+
+	if existingNode.Spec.ExternalID == node.Spec.ExternalID {
+		glog.Infof("Node %s was previously registered", kl.nodeName)
+
+		// Edge case: the node was previously registered; reconcile
+		// the value of the controller-managed attach-detach
+		// annotation.
+		requiresUpdate := kl.reconcileCMADAnnotationWithExistingNode(node, existingNode)
+		if requiresUpdate {
+			if _, err := kl.kubeClient.Core().Nodes().Update(existingNode); err != nil {
+				glog.Errorf("Unable to reconcile node %q with API server: error updating node: %v", kl.nodeName, err)
+				return false
 			}
-			if currentNode.Spec.ExternalID == node.Spec.ExternalID {
+		}
+
+		return true
+	}
+
+	if kl.cloud == nil {
```

Comment:
this is a carry we have (https://github.com/openshift/origin/commit/88abe47b9a963b41acd0c4f0fd18827192648468). We should keep the carry distinct from the upstream pick commit